### PR TITLE
#3 created user-related entities

### PR DESCRIPTION
**Pull request by MaryR-code**
Created user-related entities:
1. UserEntity
2. RoleEntity

### CODE REVIEW

File: src/main/java/guru/bug/courses/srs/entity/RoleEntity.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+@Entity
+@Table(name = "roles")
+public class RoleEntity implements Serializable {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    @Column(name="id", insertable = false, updatable = false, nullable = false)
+    private UUID id;
+
+    @Column(name="name", length = 50, nullable = false)
+    private String name;
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/entity/RoleEntity.java:17]: It isn't defined on diagram (my bad), but role's name should be unique.

File: src/main/java/guru/bug/courses/srs/entity/RoleEntity.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import javax.persistence.*;
+import java.io.Serializable;
+import java.util.Objects;
+import java.util.UUID;
+
+@Entity
+@Table(name = "roles")
+public class RoleEntity implements Serializable {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/entity/RoleEntity.java:12]: UUID won't be autogenerated. The idea of UUID that we can set it even before inserting to DB

File: src/main/java/guru/bug/courses/srs/entity/UserEntity.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+import java.io.Serializable;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+
+@Entity
+@Table(name = "users")
+public class UserEntity implements Serializable {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/entity/UserEntity.java:13]: UUID is not autoincrementable

File: src/main/java/guru/bug/courses/srs/entity/UserEntity.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    private byte[] salt;
+
+    @Column(name = "password_hash", nullable = false)
+    private byte[] passwordHash;
+
+    @ManyToMany
+    @JoinTable(name = "user_roles",
+            joinColumns = @JoinColumn(name = "user_id", referencedColumnName = "id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id", referencedColumnName = "id")
+    )
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/entity/UserEntity.java:42]: Good job! :)

File: src/main/java/guru/bug/courses/srs/entity/UserEntity.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    private List<RoleEntity> roles;
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/entity/UserEntity.java:43]: Roles are unique per user. There are no cases when user is "admin" twice. So better to migrate it to Set.
It isn't a bug - everything will work correctly even with List. More like "ideological" thing.

**Review by dima-fedotovs — CHANGES_REQUESTED** [review]
Good code style, formatting. Only thing need to be improved - ids cannot be autogenerated here.
Optionally List of Roles could be changed to Set of Roles.

File: src/main/java/guru/bug/courses/srs/entity/RoleEntity.java (hunk 1 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/entity/RoleEntity.java:16]: added

File: src/main/java/guru/bug/courses/srs/entity/UserEntity.java (hunk 4 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/entity/UserEntity.java:41]: thanks!

File: src/main/java/guru/bug/courses/srs/entity/UserEntity.java (hunk 5 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/entity/UserEntity.java]: changed

File: src/main/java/guru/bug/courses/srs/entity/RoleEntity.java (hunk 2 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/entity/RoleEntity.java]: I hope it's fixed too

**Review by yekaterinasavelyeva — APPROVED** [review]

File: src/main/java/guru/bug/courses/srs/entity/UserEntity.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+import java.io.Serializable;
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+
+@Entity
+@Table(name = "users")
+public class UserEntity implements Serializable {
+    @Id
+    @Column(name = "id", insertable = false, updatable = false, nullable = false)
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/entity/UserEntity.java:13]: the same as above

File: src/main/java/guru/bug/courses/srs/entity/RoleEntity.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import javax.persistence.*;
+import java.io.Serializable;
+import java.util.Objects;
+import java.util.UUID;
+
+@Entity
+@Table(name = "roles")
+public class RoleEntity implements Serializable {
+    @Id
+    @Column(name="id", insertable = false, updatable = false, nullable = false)
```
Comment (dima-fedotovs) [inline src/main/java/guru/bug/courses/srs/entity/RoleEntity.java:12]: Must be insertable. Updatable could be false (I have never tried this, but should be ok)
Nullable = false isn't required, as it is assumed with @Id annotation

**Review by dima-fedotovs — CHANGES_REQUESTED** [review]
insertable = false - was a bad choice. Hibernate will try to insert NULL to the corresponding column, so never will work.

File: src/main/java/guru/bug/courses/srs/entity/UserEntity.java (hunk 6 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/entity/UserEntity.java]: done!

File: src/main/java/guru/bug/courses/srs/entity/RoleEntity.java (hunk 7 above)
Comment (MaryR-code) [reply src/main/java/guru/bug/courses/srs/entity/RoleEntity.java]: done!

**Review by dima-fedotovs — APPROVED** [review]
I'm out of comments. Seems this PR can be merged.

**Pull request merged**